### DISCUSSION TITLE
Preventing runaway builds caused by multiple build triggers

### DISCUSSION THREAD

**Discussion by s-h-a-d-o-w** [caprover/caprover#2197] [Q&A] [answered]
I recently pushed a few times in quick succession to a project with a deployment GitHub action. And it seems like that was the reason why the server disk filled to 100% (it's usually only at about 15/30 GB) and CPU got stuck at near 100% (caused by rsyslog though - probably because of the space problem).

I'm posting this in Q&A because I'm wondering - I can of course avoid this by making GitHub cancel previous workflows (which... actually might not do anything because even if the workflow is cancelled, CapRover still keeps building) but... couldn't/shouldn't CapRover detect that there's already a build in progress for the same app and either wait for that to finish or reject subsequent incoming build triggers?

(If other people stumble across this: The immediate emergency can be resolved by doing `docker system prune`, then going to the CapRover UI and removing unused images.)

**Comment by githubsaturn** [comment]
This is exactly how it's designed to work.

How are you deploying your apps? Are you building the docker image on Github and just send the image address to CapRover or do you send the code to CapRover server and have it build there?


https://github.com/caprover/caprover/blob/d0bdd7ff149313bd50b14be62a5b0e92a6ae05ad/src/user/ServiceManager.ts#L115-L130

**Comment by s-h-a-d-o-w** [comment]
Building the bundle happens on GitHub servers but the docker image is built by CapRover.

It's actually a public project if you want to skim it:
https://github.com/s-h-a-d-o-w/next-simple-chatgpt/blob/master/Dockerfile
https://github.com/s-h-a-d-o-w/next-simple-chatgpt/blob/master/deploy.sh

Deployments as reported in the CapRover UI when it all started:
```
12	11/18/2024, 11:51 PM		
11	11/18/2024, 11:50 PM
```

**Comment by githubsaturn** [comment, marked as answer]
Thanks! Yes, this should work just fine based on the logic that I copied above. It is possible that your app is just a heavy app during the build process. I highly recommend building the docker image on Github.

https://github.com/caprover/deploy-from-github?tab=readme-ov-file#example-1---deploy-using-image
